File: java/algo/KnuthMorrisPratt.java
package algo;

import java.util.Arrays;

public class KnuthMorrisPratt {
    public static void main(String[] args) {
        String string = "ABCABAB ABABABAABAC";
        String pattern = "ABABAABA";

        int[] lookupTable = buildLookupTable(pattern);
        System.out.println("Lookup Table: " + Arrays.toString(lookupTable));

        stringSearch(string, pattern, lookupTable);
    }

    // Method to build lookup table using pattern
    private static int[] buildLookupTable(String pattern) {
        int[] lookupTable = new int[pattern.length()];
        Arrays.setAll(lookupTable, x -> 0);
        int idx = 0;
        int j = 1;
        while (j < pattern.length()) {
            if (pattern.charAt(idx) == pattern.charAt(j)) {
                idx++;
                lookupTable[j] = idx;
                j++;
            } else if (idx > 0) {
                idx = lookupTable[idx - 1];
            } else {
                j++;
            }
        }
        return lookupTable;
    }

    // KMP string search method
    private static void stringSearch(String string, String pattern, int[] lookupTable) {
        int i = 0, j = 0;
        while(i < string.length() && j < pattern.length()) {
            if (string.charAt(i) == pattern.charAt(j)) {
                i++;
                j++;
            } else if (j != 0) {
                j = lookupTable[j - 1];
            } else {
                i++;
            }
        }
        if (j == pattern.length()) {
            System.out.println("Found pattern at index: " + (i - j));
        } else {
            System.out.println("Could not find pattern");
        }
    }
}

File: python/knuth_morris_pratt.py
# KMP string search method
def string_search(string, pattern, lookup_table):
    i = 0
    j = 0
    while i < len(string) and j < len(pattern):
        if string[i] == pattern[j]:
            i += 1
            j += 1
        elif j > 0:
            j = lookup_table[j - 1]
        else:
            i += 1
    if j == len(pattern):
        print("Found pattern at index: ", i - j)


# Method to build lookup table using pattern
def build_lookup_table(pattern):
    table = [0 for i in range(len(pattern))]
    idx = 0
    j = 1
    while j < len(pattern):
        if pattern[idx] == pattern[j]:
            idx += 1
            table[j] = idx
            j += 1
        elif idx > 0:
            idx = table[idx - 1]
        else:
            j += 1
    return table


# Driver method
def main():
    string = "ABCABAB ABABABAABAC"
    pattern = "ABABAABA"
    lookup_table = build_lookup_table(pattern)
    print("Lookup table: ", lookup_table)
    string_search(string, pattern, lookup_table)


if __name__ == "__main__":
    main()
